apps/passport-client/src/validateState.ts

```diff
@@ -180,113 +173,95 @@ export function getRunningAppStateValidationErrors(
-      );
-    }
-    if (commitmentFromSelfField !== commitmentFromIdentityField) {
-      errors.push(
-        `commitment in 'self' field of app state (${commitmentFromSelfField})` +
-          ` does not match commitment of 'identity' field of app state (${commitmentFromIdentityField})`
-      );
-    }
-
-    if (commitmentFromIdentityField !== commitmentOfIdentityPCDInCollection) {
+    if (
+      v3CommitmentFromIdentityField !== v3CommitmentOfIdentityPCDInCollection
+    ) {
       errors.push(
-        `commitment of 'identity' field of app state (${commitmentFromIdentityField})` +
-          ` does not match commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})`
+        `commitment of 'identity' field of app state (${v3CommitmentFromIdentityField})` +
+          ` does not match commitment of identity pcd in collection (${v3CommitmentOfIdentityPCDInCollection})`
       );
     }
   }
 
-  if (
-    commitmentV4FromSelfField === undefined ||
-    commitmentOfIdentityV4PCDInCollection === undefined ||
-    commitmentV4FromIdentityField === undefined
-  ) {
-    // these cases are validated earlier in this function
-  } else {
-    // in 'else' block we check that the commitments from all three
-    // places that the user's commitment exists match - in the self, the
-    // identity, and in the pcd collection
+  const v4CommitmentFromPCDCollection =
+    identityPCDFromCollection?.claim.identityV4?.commitment?.toString();
+  const v4CommitmentFromSelfField = self?.semaphore_v4_commitment;
 
-    if (commitmentOfIdentityV4PCDInCollection !== commitmentV4FromSelfField) {
-      errors.push(
-        `commitment of identity v4 pcd in collection (${commitmentOfIdentityV4PCDInCollection})` +
-          ` does not match commitment in 'self' field of app state (${commitmentV4FromSelfField})`
-      );
-    }
-    if (commitmentV4FromSelfField !== commitmentV4FromIdentityField) {
+  const v4PublicKeyFromSelfField = self?.semaphore_v4_pubkey;
+  const v4PublicKeyFromPCDCollection =
+    identityPCDFromCollection &&
+    v4PublicKey(identityPCDFromCollection.claim.identityV4);
+
+  // either we are missing v4 commitment from `self`
+  if (!v4CommitmentFromSelfField && !!v4PublicKeyFromSelfField) {
+    errors.push(
+      "missing 'semaphore_v4_commitment' from 'self'. either both 'semaphore_v4_commitment' and 'semaphore_v4_pubkey' must be present, or neither must be present."
+    );
+  }
+  // or we are missing v4 public key from `self`
+  else if (!!v4CommitmentFromSelfField && !v4PublicKeyFromSelfField) {
+    errors.push(
+      "missing 'semaphore_v4_pubkey' from 'self'. either both 'semaphore_v4_commitment' and 'semaphore_v4_pubkey' must be present, or neither must be present."
+    );
+  }
+  // or both v4 public key and v4 commitment are present in `self`
+  else if (v4CommitmentFromSelfField && v4PublicKeyFromSelfField) {
```

Comment:
Why no "!!" on these, compared to the corresponding checks above?

Reply:
no particular reason - the behavior is the same between the two cases, including the `!!` is just a bit more explicit and intentional.